test/test_common/simulated_time_system.cc

```diff
@@ -47,15 +47,163 @@ class UnlockGuard {
+        // and ensures that events scheduled from outside the worker thread have a chance to
+        // execute. It may be necessary to have the simulated scheduler add a real periodic timer on
+        // construction in order to ensure that the event loop remains responsive to external event
+        // activations.
+        inc_pending = true;
+        pending_dec_ = true;
+      }
+    }
+    if (inc_pending) {
+      time_system_.incPending();
+    }
+    if (!update_time_cb_->enabled()) {
+      run_alarms_cb_->cancel();
+      update_time_cb_->scheduleCallbackCurrentIteration();
+    }
+  }
+
+private:
+  void disableAlarmLockHeld(Alarm& alarm) ABSL_EXCLUSIVE_LOCKS_REQUIRED(mutex_);
+
+  void updateTimeAndTriggerAlarms();
+  void runReadyAlarms();
+
+  struct AlarmRegistration {
+    AlarmRegistration(MonotonicTime time, uint64_t randomness, Alarm& alarm)
+        : time_(time), randomness_(randomness), alarm_(alarm) {}
+
+    MonotonicTime time_;
+    // Random tie-breaker for alarms scheduled for the same monotonic time used to mimic
+    // non-deterministic execution of real alarms scheduled for the same wall time.
+    uint64_t randomness_;
+    Alarm& alarm_;
+
+    friend bool operator<(const AlarmRegistration& lhs, const AlarmRegistration& rhs) {
+      if (lhs.time_ != rhs.time_) {
+        return lhs.time_ < rhs.time_;
+      }
+      if (lhs.randomness_ != rhs.randomness_) {
+        return lhs.randomness_ < rhs.randomness_;
+      }
+      // Out of paranoia, use pointer comparison on the alarms as a final tie-breaker but also
+      // ASSERT that this branch isn't hit in debug modes since in practice the randomness_
+      // associated with two registrations should never be equal.
+      ASSERT(false, "Alarm registration randomness_ for two alarms should never be equal.");
+      return &lhs.alarm_ < &rhs.alarm_;
+    }
+  };
+
+  class AlarmSet {
+  public:
+    bool empty() { return alarms_.empty(); }
+
+    const AlarmRegistration& next() { return *alarms_.begin(); }
+
+    void add(AlarmRegistration registration) {
+      auto insert_result = alarms_.insert(registration);
+      ASSERT(insert_result.second);
+      alarm_registrations_map_.emplace(&registration.alarm_, insert_result.first);
+
+      // Sanity check that the parallel data structures used for alarm registration have the same
+      // number of entries.
+      ASSERT(alarms_.size() == alarm_registrations_map_.size());
+    }
+
+    bool remove(Alarm& alarm) {
+      auto it = alarm_registrations_map_.find(&alarm);
+      if (it == alarm_registrations_map_.end()) {
+        return false;
+      }
+      alarms_.erase(it->second);
+      alarm_registrations_map_.erase(it);
+      return true;
+    }
+
+    bool contains(Alarm& alarm) const {
+      return alarm_registrations_map_.find(&alarm) != alarm_registrations_map_.end();
+    }
+
+  private:
+    std::set<AlarmRegistration> alarms_;
```

Comment:
nit (this was pre-existing): call this ordered_alarms_, just for clarity at use?